lib/output_helper.rb

```diff
@@ -90,6 +90,190 @@ def self.csv_line(vrp, service, cluster_index, two_stages = false)
     end
   end
 
+  class Result
+    def self.generate_geometry(expected_geometry, results)
+      @colors = ['#9000EE', '#00CC00', '#009EFF', '#0077A3', '#000000', '#CC1882',
+                 '#003880', '#DD0000', '#558800', '#FFBB00', '#BEE562']
+      @cluster_colors = {}
+
+      expected_geometry.map!(&:to_sym)
+      results.each{ |result|
+        result[:geojsons] = [{}]
+        @vehicle_partitioned = false
+        @work_day_partitioned = false
+
+        generate_partitions_geometry(result) if expected_geometry.include?(:partitions)
+        generate_points_geometry(result)
+        generate_polylines_geometry(result) if expected_geometry.include?(:polylines)
+      }
+    end
+
+    def self.generate_partitions_geometry(result)
+      activities = result[:routes].collect{ |r|
+        r[:activities].select{ |a| a[:type] != 'depot' && a[:type] != 'rest' }
+      }.flatten
+      elements = activities + result[:unassigned]
+      all_skills = elements.collect{ |a| a[:detail][:internal_skills] }.flatten.uniq
+
+      has_vehicle_partition = all_skills.any?{ |skill| skill.to_s.include?('vehicle_partition_') }
+      has_work_day_partition = all_skills.any?{ |skill| skill.to_s.include?('work_day_partition_') }
+      return unless has_vehicle_partition || has_work_day_partition
+
+      result[:geojsons][0][:partitions] ||= {}
+      if has_vehicle_partition
+        @vehicle_partitioned = true
+        result[:geojsons][0][:partitions][:vehicle] = draw_cluster(elements, :vehicle)
+      end
+
+      return unless has_work_day_partition
+
+      @work_day_partitioned = true
+      result[:geojsons][0][:partitions][:work_day] = draw_cluster(elements, :work_day)
+    end
+
+    def self.draw_cluster(elements, entity)
+      polygons = elements.group_by{ |element|
+        entity == :vehicle ?
+                  [element[:detail][:internal_skills].find{ |sk| sk.to_s.include?('vehicle_partition_') }] :
+                  [element[:detail][:internal_skills].find{ |sk| sk.to_s.include?('work_day_partition_') },
+                   element[:detail][:skills].find{ |sk| sk.include?('cluster ') }]
```

Comment:
We need _find_ because internal skills contain all service skills. Therefore we can not group_by skills/internal_skills but we need to select interesting skills.

We had decided that internal skills are original_skills were we add more data (instead of generating specific structure for clustering skills).